tests/convective_op/convective_op.cpp

```diff
@@ -0,0 +1,319 @@
+        }
+
+        gridding_algorithm->makeCoarsestLevel(patch_hierarchy, 0.0);
+        int tag_buffer = 1;
+        int level_number = 0;
+        bool done = false;
+        while (!done && (gridding_algorithm->levelCanBeRefined(level_number)))
+        {
+            gridding_algorithm->makeFinerLevel(patch_hierarchy, 0.0, 0.0, tag_buffer);
+            done = !patch_hierarchy->finerLevelExists(level_number);
+            ++level_number;
+        }
+
+        // Allocate data on each level of the patch hierarchy.
+        const int coarsest_ln = 0;
+        const int finest_ln = patch_hierarchy->getFinestLevelNumber();
+        allocate_patch_data({ un_idx,
+                              us_idx,
+                              N_un_idx,
+                              N_us_idx,
+                              e_un_idx,
+                              e_us_idx,
+                              thn_idx,
+                              draw_un_idx,
+                              draw_us_idx,
+                              draw_en_idx,
+                              draw_es_idx,
+                              draw_Nn_idx,
+                              draw_Ns_idx,
+                              exact_N_un_idx,
+                              exact_N_us_idx },
+                            patch_hierarchy,
+                            0.0,
+                            coarsest_ln,
+                            finest_ln);
+
+        // Setup velocity and pressures functions.
+        muParserCartGridFunction un_fcn("un", app_initializer->getComponentDatabase("un"), grid_geometry);
+        muParserCartGridFunction us_fcn("us", app_initializer->getComponentDatabase("us"), grid_geometry);
+
+        // Set up Thn functions
+        muParserCartGridFunction thn_fcn("thn", app_initializer->getComponentDatabase("thn"), grid_geometry);
+
+        // Setup exact solution functions
+        muParserCartGridFunction N_un_fcn("N_un", app_initializer->getComponentDatabase("N_un"), grid_geometry);
+        muParserCartGridFunction N_us_fcn("N_us", app_initializer->getComponentDatabase("N_us"), grid_geometry);
+
+        un_fcn.setDataOnPatchHierarchy(un_idx, un_var, patch_hierarchy, 0.0);
+        us_fcn.setDataOnPatchHierarchy(us_idx, us_var, patch_hierarchy, 0.0);
+        thn_fcn.setDataOnPatchHierarchy(thn_idx, thn_var, patch_hierarchy, 0.0);
+
+        N_un_fcn.setDataOnPatchHierarchy(e_un_idx, e_un_var, patch_hierarchy, 0.0);
+        N_us_fcn.setDataOnPatchHierarchy(e_us_idx, e_us_var, patch_hierarchy, 0.0);
+
+        // Apply convective operator
+        INSVCTwoFluidConvectiveManager convec_op("convec_op",
+                                                 patch_hierarchy,
+                                                 input_db->getDatabase("ConvecOp"),
+                                                 un_bc_coefs,
+                                                 us_bc_coefs,
+                                                 thn_bc_coef.get());
```

Comment:
The issue is here. The patch index with the largest ghost cell width needs to be created before the patch hierarchy is created. This line just needs to be moved before line 178

Reply:
I tried that last night, but still ran into the same error. Let me check again though...

Reply:
Results in a seg fault for both uniform and amr grids:

```
Thread 1 "convective_op_2" received signal SIGSEGV, Segmentation fault.
0x00005555555c749a in SAMRAI::tbox::Pointer<SAMRAI::hier::BasePatchLevel<2> >::Pointer(SAMRAI::tbox::PointerBase const&) ()
```

Reply:
I really hate samrai sometimes. They make it so hard to do RAII.

The issue is some of the members of `INSVCTwoFluidConvectiveManager` rely on the patch hierarchy being fully functional. In particular, `d_hier_sc_data_ops` and `d_hier_cc_data_ops` should be replaced with `std::unique_ptr<>` and be created as needed, probably in `allocateData()`.

Reply:
Can you do this? I can also push a change to this branch if you want me to.

Reply:
Yeah, let me try this